homeassistant/components/remootio/config_flow.py

```diff
@@ -0,0 +1,184 @@
+            vol.Coerce(str),
+            vol.Upper,
+            vol.Match(REGEX_CREDENTIAL),
+            msg="API Auth Key appears to be invalid",
+        ),
+        vol.Required(
+            CONF_DEVICE_CLASS,
+            default=DEVICE_CLASS_GARAGE,
+            msg="Controlled device's class is required",
+        ): vol.All(
+            vol.Coerce(str),
+            vol.In([DEVICE_CLASS_GARAGE, DEVICE_CLASS_GATE]),
+            msg="Controlled device's class appears to be invalid",
+        ),
+    },
+    extra=REMOVE_EXTRA,
+)
+
+DEVICE_NAME = "Remootio Device"
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input."""
+
+    _LOGGER.debug("Validating input... Input [%s]", data)
+
+    INPUT_VALIDATION_SCHEMA(data)
+
+    connection_options: ConnectionOptions = ConnectionOptions(
+        data[CONF_IP_ADDRESS], data[CONF_API_SECRET_KEY], data[CONF_API_AUTH_KEY]
+    )
+
+    device_serial_number: str = await get_serial_number(
+        hass, connection_options, _LOGGER
+    )
+
+    return {
+        CONF_TITLE: f"{DEVICE_NAME} (IP: {data[CONF_IP_ADDRESS]}, S/N: {device_serial_number})",
+        CONF_SERIAL_NUMBER: device_serial_number,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Remootio."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the class."""
+        super().__init__()
+
+        self.form_displayed = False
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step."""
+        user_input = user_input or {}
+        errors = {}
+
+        if user_input is not None and len(user_input) != 0:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except UnsupportedRemootioDeviceError:
+                _LOGGER.debug("Remootio device isn't supported.", exc_info=True)
+                return self.async_abort(reason="unsupported_device")
+            except vol.MultipleInvalid as e:
+                incomplete_data = True
+
+                _LOGGER.debug(
+                    f"Invalid user input. MultipleInvalid.Errors [{e.errors}]",
+                    exc_info=True,
+                )
+                for error in e.errors:
+                    _LOGGER.debug(
+                        f"Error [{error.__class__.__name__}] Path [{error.path[0]}]"
+                    )
+                    if isinstance(error, RequiredFieldInvalid):
+                        errors[str(error.path[0])] = f"{error.path[0]}_required"
+                    else:
+                        errors[str(error.path[0])] = f"{error.path[0]}_invalid"
+                        incomplete_data = False
+
+                if incomplete_data and not self.form_displayed:
+                    errors = {}
+                    errors["base"] = "user_input_incomplete"
```

Comment:
> This overwrite any required field invalid errors from above.

This is intentional, because the device can also be discovered using zeroconf, but this will be implemented in a later version of the integration.

If the device is discovered using zeroconf, only the IP address is known first, the other data (API Secret Key and API Auth Key) must be given by the user. The idea was that in this case only the error message _Please complete the setup of your Remootio device._ with key `user_input_incomplete` would be presented to the user.

Reply:
Seems a bit confusing. You can have different flows.

Reply:
I intend to reuse the method again, as for example with the integration for Gogogate2.

Reply:
You can't reuse this file. This is specific to this integration. If you want to keep it around keep a local copy 

If you need that feature later code related to that should be added at that stage.

Reply:
I mean that Remootio devices can be discovered using zeroconf. I want to implement this in a later version of the integration and then reuse/call this method to let the user give the API Secret Key, the API Auth Key and the device class.

Reply:
While i think it make more sense with a set of helpers used in both flows, i don't see how your want to resue this function affects things. Should still be possible.
That said, we refrain from preparing for future changes if the code can be written simpler initially. Reviews become harder, code more complex initially, and you might end up wanting to do things differently later.

Reply:
I'll adjust it.

Reply:
Fixed.